### PR TITLE
etcdwatch: clean up watch code

### LINKED ISSUE #86: Make watch API stable
- [ ] Revisit watch API to make it consistent with etcd/clientv3 Watch API
- [ ] Add test

### PR DESCRIPTION
**Pull request by fanminshi**
Fixes #86

### CODE REVIEW

**Comment by fanminshi** [comment]
Working in progress.

**Comment by riclava** [comment]
why travis ci build failed?

**Comment by fanminshi** [comment]
@riclava my unit test for watch is failing. need to take a look.

File: src/main/java/com/coreos/jetcd/EtcdWatch.java [hunk 1]
```diff
@@ -1,74 +1,81 @@
...
-    void onWatch(EtcdHeader header, List<WatchEvent> events);
-
-    /**
-     * onResuming will be called when the watcher is on resuming.
+     * @throws InterruptedException if interrupted while waiting.
+     * @throws ConnectException if watch client experiences unrecoverable connection error.
+     * @throws TimeoutException if watcher timeout on poll.
      */
-    void onResuming();
+    WatchResponse poll(long timeout, TimeUnit unit)
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/EtcdWatch.java]: how about removing take? only keep poll?

also poll only takes one parameter as duration in nanoseconds?

Comment (ncubrian) [inline src/main/java/com/coreos/jetcd/EtcdWatch.java]: Could you consider adding withPrefix option to WatchOption?

Comment (xiang90) [inline src/main/java/com/coreos/jetcd/EtcdWatch.java]: that is when you create the watch, not when you do the polling.

Comment (ncubrian) [inline src/main/java/com/coreos/jetcd/EtcdWatch.java]: @xiang90 I know. I'm just wondering if you have a plan on adding watch with prefix.

Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/EtcdWatch.java]: @ncubrian I will add watch with prefix option. Also currently, watch do support prefix behavior. you can use `withRange(ByteSequence endKey)` where endkey is one bit larger than the given key, then it gets all keys with the prefix.

File: src/main/java/com/coreos/jetcd/exception/CompactedException.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package com.coreos.jetcd.exception;
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/exception/CompactedException.java]: can we add etcd exception in another pr?

Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/exception/CompactedException.java]: yeah I can.

**Comment by xiang90** [comment]
actually the first few prs can be separated from this one. can you also do that. keeping each pr small is the goal.

**Comment by fanminshi** [comment]
I have clean up the watch. PTAL.

**Comment by fanminshi** [comment]
cleaned up this pr. ptal.

File: src/main/java/com/coreos/jetcd/data/Header.java [hunk 3]
```diff
@@ -9,15 +9,12 @@
   private final long memberId;
   private final long revision;
   private final long raftTerm;
-  private final long compactRevision;
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/data/Header.java]: can you also move this commit to a separate pr?

File: src/main/java/com/coreos/jetcd/internal/impl/Util.java [hunk 4]
```diff
@@ -172,7 +173,12 @@ static DeleteResponse toDeleteResponse(DeleteRangeResponse rangeResponse) {
    * convert API CompactionResponse to client CompactResponse.
    */
   static CompactResponse toCompactResponse(CompactionResponse response) {
-    return new CompactResponse(toHeader(response.getHeader(), 0));
+    return new CompactResponse(toHeader(response.getHeader()));
+  }
+
+  static WatchResponse toWatchResponse(com.coreos.jetcd.api.WatchResponse response) {
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/internal/impl/Util.java]: also this commit should be in a separate pr.

File: src/main/java/com/coreos/jetcd/watch/WatchCreateException.java [hunk 5]
```diff
@@ -1,16 +0,0 @@
-package com.coreos.jetcd.watch;
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/watch/WatchCreateException.java]: this should be in a separate pr.

File: src/main/java/com/coreos/jetcd/options/WatchOption.java [hunk 6]
```diff
@@ -98,11 +97,6 @@ public Builder withNoPut(boolean noPut) {
       return this;
     }
 
-    public Builder withResuming(boolean resuming) {
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/options/WatchOption.java]: this can be done in a separate pr

File: src/main/java/com/coreos/jetcd/watch/WatchCreateException.java (hunk 5 above)
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/watch/WatchCreateException.java]: WatchCreateException is used in the current(not this pr) watch impl. It will break the ci if i just delete this exception.

File: src/main/java/com/coreos/jetcd/options/WatchOption.java (hunk 6 above)
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/options/WatchOption.java]: yep, can do.

File: src/main/java/com/coreos/jetcd/watch/WatchCreateException.java (hunk 5 above)
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/watch/WatchCreateException.java]: ok. then keep it here.

File: src/main/java/com/coreos/jetcd/data/Header.java (hunk 3 above)
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/data/Header.java]: header is tied to watch impl. this change will break ci. it must be changed along with watch impl.

Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/data/Header.java]: Maybe I don't change Header to include in this pr since compactRevision doesn't break anything. Once the watch pr is merged then I will have another pr to remove compactRevision.

File: src/main/java/com/coreos/jetcd/internal/impl/Util.java (hunk 4 above)
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/internal/impl/Util.java]: yep will do.

File: src/main/java/com/coreos/jetcd/options/WatchOption.java (hunk 6 above)
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/options/WatchOption.java]: take a look. it is tied with watch impl. I can't remove it alone.

File: src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java [hunk 7]
```diff
@@ -1,44 +1,69 @@
...
+  private static final Logger logger = Logger.getLogger(WatchImpl.class.getName());
 
   private final WatchGrpc.WatchStub watchStub;
 
-  private ConcurrentLinkedQueue<Pair<WatcherImpl, CompletableFuture<Watcher>>>
-      pendingCreateWatchers = new ConcurrentLinkedQueue<>();
-  private Map<Long, CompletableFuture<Boolean>> pendingCancelFutures = new ConcurrentHashMap<>();
+  private volatile StreamObserver<WatchRequest> grpcWatchStreamObserver;
+
+  private final ConcurrentHashMap<Long, WatcherImpl> watchers = new ConcurrentHashMap<>();
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: what is the key of the hash map? document it?

Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: Lease-> WatcherImpl. and will document.

File: src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java [hunk 8]
```diff
@@ -48,234 +73,275 @@
...
+    Preconditions.checkState(!isClosed(), "Watch client has been closed");
+
+    this.setClosed();
+    this.notifyWatchers(EtcdExceptionFactory.newEtcdException("Watch client has been closed"));
+    this.closeGrpcWatchStreamObserver();
+    this.executor.shutdownNow();
+    this.scheduledExecutorService.shutdownNow();
+  }
+
+  private void notifyWatchers(EtcdException e) {
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: closeAllWatchersWithError?

Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: the log message is not accurate. the client should notify the watchers with a exception but not closing it. user is responsible for closing watcher.

File: src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java [hunk 9]
```diff
@@ -48,234 +73,275 @@
...
+    return status.getCode() != Status.Code.UNAVAILABLE && status.getCode() != Status.Code.INTERNAL;
+  }
+
+  private void processCreate(WatchResponse response) {
+    WatcherImpl watcher = this.pendingWatchers.poll();
+
+    this.sendNextWatchCreateRequest();
+
+    if (watcher == null) {
+      // shouldn't happen
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: throw exception? not sure what is the best practice for java to panic stuff.

File: src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java (hunk 8 above)
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: so notifyWatchers only notifies watchers but close them.

**Comment by xiang90** [comment]
The interface is much cleaner. The impl seems ok. We can get this merged. I want to add some integration tests after this PR.

File: src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java (hunk 8 above)
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: why not close the watcher for the users?

File: src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java (hunk 9 above)
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: In the case of duplicate create response. this situation can happen but it shouldn't. I think log this case will be fine since correctness is not compromised by duplicate create messages.

File: src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java [hunk 10]
```diff
@@ -48,234 +73,275 @@
...
+    }
+    this.grpcWatchStreamObserver.onCompleted();
+    this.grpcWatchStreamObserver = null;
+  }
+
+  private boolean isNoLeaderError(Status status) {
+    if (status.getDescription() == null) {
+      return false;
+    }
+    return status.getDescription().contains("no leader");
```
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: I can build a map of etcd server error to jetcd error. then use that map to determine "no leader" error. I can do that in a separate pr.

**Comment by fanminshi** [comment]
comments addressed. PTAL.

**Comment by xiang90** [comment]
lgtm in general.
